rfd/0064-audit-modes.md

```diff
@@ -0,0 +1,190 @@
+| `node-async` and `proxy-async` | Connection errors with Proxy/Auth server while streaming audit logs/session recording data. |
+
+There are going to be two audit modes:
+
+* **Strict**: Teleport won't tolerate session recording or audit logging to fail.
+  In practice, if Teleport cannot record the session, it will not start or will
+  be terminated (in case it is already in progress). This mode will be the
+  recommended one as it guarantees all sessions are being recorded.
+* **Best-effort**: Audit failures are tolerated in this mode, meaning that Teleport
+  will allow sessions without recording them. System administrators might use
+  this mode to be able to access resources continually.
+
+## UX
+
+### Configuration
+
+The configuration is going to be done at the role level. The role option
+`record_session` will be extended to hold the audit mode values: `strict` and
+`best_effort`. It will have one entry per session kind (desktop, databases,
+applications, Kubernetes, and SSH) and an `all` entry for cases where users
+want the same behavior for all kinds.
+
+**Note**: Currently, the option  `record_session.desktop` uses a boolean value to
+disable/enable session recording. It will have to be converted into a new
+format: `true` will be charged to the default audit mode, and `false` will be
+`off`. Disabling session recording is out of the scope of this RFD, meaning that
+the `off` value will only be present on the `desktop` kind.
+
+**Role example (with all session kinds):**
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: alice
+spec:
+  allow:
+    ...
+  options:
+    record_session:
+      all: strict|best_effort
+      ssh: strict|best_effort
+      app: strict|best_effort
+      db: strict|best_effort
+      k8s: strict|best_effort
+      desktop: strict|best_effort|off
+```
+
+### Examples
+
+### Strict mode
+
+```shell
+# Prevent users from starting new SSH sessions when it encounters an audit error.
+$ tsh ssh root@node
+Node could not record audit events: no space left on device.
+ERROR: ssh: could not start shell
+```
+
+```shell
+# Terminate the SSH session if it encounters an audit error.
+$ tsh ssh root@node
+root@node:~$ echo "Hello server"
+Hello server
+
+Node could not record audit events: no space left on device.
+Closing session...
+
+# Same for Kubernetes exec sessions.
+$ tsh kube exec -t pod-name bash
+root@container:~$ echo "Hello server"
+Hello server
+
+Node could not record audit events: no space left on device.
+Closing session...
+```
+
+```shell
+# Non-interactive sessions (such as Applications and K8S) will fail the next request made.
+$ tsh app login dumper
+$ curl ...
+Forbidden error...
+
+$ tsh kube login my-cluster
+$ kubectl ...
+Forbidden error...
+```
+
+### Best effort mode
+
+```shell
+# SSH Session starts with a "warning" message.
+$ tsh ssh root@node
+Disabling recording and audit events for this session, reason: node error: no space left on device
```

Comment:
Should we give this information to the end user?
This could be leveraged by some bad actor to know when they can execute commands that won't be "supervised", or just being less cautious
I mean, this is not the default mode so the admin is already aware that this could happen anyway

Reply:
Yes, IMO would be a better option
(`s/correctly/incorrectly/`)